FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RecipientsViewModel.kt

```diff
@@ -0,0 +1,391 @@
+                  e.printStackTrace()
+                  publicKeyEntity.isNotUsable = true
+                }
+              }
+            }
+
+            try {
+              getPgpInfoFromServer(email = emailLowerCase)?.let { pgpKeyDetailsList ->
+                cachedRecipientWithPubKeys?.let { recipientWithPubKeys ->
+                  updateCachedInfoWithPubKeysFromLookUp(
+                    recipientWithPubKeys,
+                    pgpKeyDetailsList
+                  )
+                }
+              }
+              cachedRecipientWithPubKeys = getCachedRecipientWithPubKeys(emailLowerCase)
+              cachedRecipientWithPubKeys?.let { recipients.add(it) }
+            } catch (e: Exception) {
+              e.printStackTrace()
+              ExceptionUtil.handleError(e)
+            }
+          }
+        }
+        setResultForRemoteContactsLiveData(type, Result.success(recipients))
+      } catch (e: Exception) {
+        e.printStackTrace()
+        ExceptionUtil.handleError(e)
+        setResultForRemoteContactsLiveData(type, Result.exception(e))
+      }
+    }
+  }
+
+  private suspend fun getCachedRecipientWithPubKeys(emailLowerCase: String): RecipientWithPubKeys? =
+    withContext(Dispatchers.IO) {
+      val cachedRecipientWithPubKeys = roomDatabase.recipientDao()
+        .getRecipientWithPubKeysByEmailSuspend(emailLowerCase) ?: return@withContext null
+
+      for (publicKeyEntity in cachedRecipientWithPubKeys.publicKeys) {
+        try {
+          val result = PgpKey.parseKeys(publicKeyEntity.publicKey).pgpKeyDetailsList
+          publicKeyEntity.pgpKeyDetails = result.firstOrNull()
+        } catch (e: Exception) {
+          e.printStackTrace()
+          publicKeyEntity.isNotUsable = true
+        }
+      }
+      return@withContext cachedRecipientWithPubKeys
+    }
+
+  private suspend fun updateCachedInfoWithPubKeysFromLookUp(
+    cachedRecipientEntity: RecipientWithPubKeys, list: List<PgpKeyDetails>
+  ) = withContext(Dispatchers.IO) {
+    val email = cachedRecipientEntity.recipient.email
+    val mapOfUniquePubKeysFromLookUp = mutableMapOf<String, PgpKeyDetails>()
+
+    for (pgpKeyDetails in list) {
+      val fingerprint = pgpKeyDetails.fingerprint
+      val existingPgpKeyDetails = mapOfUniquePubKeysFromLookUp[fingerprint]
+      if (existingPgpKeyDetails == null) {
+        mapOfUniquePubKeysFromLookUp[fingerprint] = pgpKeyDetails
+      } else {
+        val existingLastModified = existingPgpKeyDetails.lastModified ?: 0
+        val candidateLastModified = pgpKeyDetails.lastModified ?: 0
+        if (candidateLastModified > existingLastModified) {
+          mapOfUniquePubKeysFromLookUp[fingerprint] = pgpKeyDetails
+        }
+      }
+    }
+
+    val lookUpList = mapOfUniquePubKeysFromLookUp.values
+    for (pgpKeyDetails in lookUpList) {
+      val existingPublicKeyEntity =
+        cachedRecipientEntity.publicKeys.firstOrNull { it.fingerprint == pgpKeyDetails.fingerprint }
```

Comment:
Once we rename the `existing` above to `alreadyEncountered`, then the `existing` down here will have more meaning to it.